src/lib.rs

```diff
@@ -1,145 +1,169 @@
+    pub port: PortGroup,
+    // An array of LED structs
+    pub led: Vec<LED>,
+    // A single button struct
+    pub button: Button
 }
 
-impl TesselPort {
-
-  pub fn new(p: &'static str) -> Result<TesselPort, &'static str> {
-    match p {
-      "a" | "A" => Ok(TesselPort {
-        sock: UnixStream::connect(&Path::new(USD_PORT_A)).unwrap()
-      }),
-      "b" | "B" => Ok(TesselPort {
-        sock: UnixStream::connect(&Path::new(USD_PORT_B)).unwrap()
-      }),
-      _ => Err("Invalid port selected. Please choose either \"a\" or \"b\""),
+impl Tessel {
+    // Factory metho for Tessel
+    pub fn new() -> Tessel {
+
+        // Create a port group with two ports, one on each domain socket path
+        let ports = PortGroup { a: Port::new(PORT_A_UDS_PATH), b: Port::new(PORT_B_UDS_PATH) };
+
+        // Create models for the four LEDs
+        let red_led = LED::new("red", "error");
+        let amber_led = LED::new("amber", "wlan");
+        let green_led = LED::new("green", "user1");
+        let blue_led = LED::new("blue", "user2");
+
+        // Create the button
+        let button =  Button::new();
+
+        // Return the Tessel with these fields
+        Tessel {
+            port: ports,
+            led: vec![red_led, amber_led, green_led, blue_led],
+            button: button
+        }
     }
-  }
+}
 
-  pub fn run(&mut self, actions: &mut[Action]) -> IOResult<()> {
-    for i in actions.iter() {
-      try!(self.sock.write_all(&[i.cmd, i.arg]));
-      try!(self.sock.write_all(i.tx));
+// A group is a simple way to access each port through its letter identifier
+#[allow(dead_code)]
+pub struct PortGroup {
+    a: Port,
+    b: Port
+}
+
+// A model of the Tessel hardware ports
+#[allow(dead_code)]
+pub struct Port {
+    // string slice to the path of the domain socket
+    socket_path: &'static str
+}
+
+impl Port {
+    // Factory method for returning a new port struct
+    pub fn new(path: &'static str) -> Port {
+        Port {
+            socket_path: path
+        }
     }
+}
+
+#[allow(dead_code)]
+pub struct LED {
+    // The color of the given LED (used in filepath creation)
+    color: &'static str,
+    // The type of LED (used in filepath creation)
+    kind: &'static str,
+    // The file object we write to in order to change state
+    file: File,
+    // The current value of the LED, defaults to false
+    value: bool
+}
 
-    for i in actions.iter_mut() {
-      // TODO - get an equivalent of read_at_least in here
-      try!(self.read_all(i.rx));
+// Prefix and suffix to every LED file path
+const LED_PATH_PREFIX: &'static str = "/sys/devices/leds/leds/tessel:";
+const LED_PATH_SUFFIX: &'static str = "/brightness";
+impl LED {
+    // Factory method for creating new LEDs
+    pub fn new(color: &'static str, kind: &'static str) -> LED {
+
+        // Assemble the file path
+        let name = color.to_string() + ":" + kind;
+        let path = LED_PATH_PREFIX.to_string() + &name + LED_PATH_SUFFIX;
```

Comment:
`format!("/sys/device/leds/leds/tessel:{}:{}/brightness", color, kind);` could be easier.